### PR TITLE
Bootstrap Companions in Kubernetes Environments

### LINKED ISSUE #123: Persistent Data
At the moment PREvant does not provide any option to ensure that the data of a service or companion will be persisted. For example, if someone operates a PostgreSQL database on a Kubernetes cluster and the container will be moved to another node, all data will be lost for this application (Kafka is another example).

Further, people report to me personally, that they need a way of sharing data between different services or companions (even when [Microservices shouldn't be synchronized via file system because it is the worst option](https://youtu.be/bbotWkCLuF0?t=355))

Anyways, PREvant should allow that services or companions are able to store data. Here are some ideas:

- The `registry.rs` can be enhanced not just returning the port but also including the list of exposed volumes (according to the [spec](https://github.com/moby/moby/blob/a30990b3c8d0d42280fa501287859e1d2393a951/image/spec/v1.md#image-json-description)
- If configured, PREvant can create volumes in the case of Docker backend and PVCs in the case of Kubernetes backend for companions automatically.
  - What kind of PVC should be configurable for Kubernetes backend
- Hooks should provide a mechanism to share persistent data

### PR DESCRIPTION
**Pull request by schrieveslaach**
This commit introduces the configuration `bootstrapping.containers` which provides a way to parse the configuration of application wide companions because the current available configuration of companions is quite limiting (Current backends, Docker and Kubernetes, offer way more options than the PREvant configuration object allows). For example, PREvant was limited to self-contained applications where each microservice only relies on interactions via network API calls (REST, database connections, messaging, etc.) With this commit PREvant is now able to deploy application companions that are more powerful than the PREvant configuration in Kubernetes backends.

If `bootstrapping.containers` is defined, PREvant will start one or more containers on the infrastructure backend that are expected to generate Kubernetes manifests as output on standard out (stdout) that will be parsed by PREvant and supported are:

 - roles and role bindings
 - config maps and secrets
 - service accounts
 - persistent volume claims
 - services
 - pods, deployments, stateful sets, and jobs

Then before deploying these manifests PREvant merges all objects with the objects generated from the HTTP request payload. Thus you can add or overwrite configurations. For example, you can change the image used or an environment variable. If you overwrite any configuration the companion will be turned into an instance (as PREvant did before).

Ingresses won't be deployed if the bootstrap container outputs one of these. Instead they will be parsed and if they use the ingress class `nginx` they will be transformed into Traefik ingresses and middlewares so that the microservices will be available via web interface.

This approach make #143 obsolete and fixes #123 and contributes to #146.

### CODE REVIEW

File: docs/companions.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
+# Companion Configuration
+
+Have a look at the [basic terminology](../README.md) what is a companion. For
+these use cases following sections provide example configurations.
```
Comment (aixigo-ohornischer) [inline docs/companions.md:4]:
```suggestion
these use cases the following sections provide example configurations.
```

File: docs/companions.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
+# Companion Configuration
+
+Have a look at the [basic terminology](../README.md) what is a companion. For
```
Comment (aixigo-ohornischer) [inline docs/companions.md:3]:
```suggestion
Have a look at the [basic terminology](../README.md) to understand what a companion is. For
```

File: docs/companions.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+these use cases following sections provide example configurations.
+
+A simple but limited configuration of companions can be done via the
```
Comment (aixigo-ohornischer) [inline docs/companions.md:6]:
```suggestion
A simple, but limited configuration of companions can be done via the
```

File: docs/companions.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+`config.toml` file for [application companions](#application-wide) and [service
+companions](#service-based). More complex companions can be created via
+[bootstrapping](#bootstrapping-from-the-infrastructure-backend).
+
+## Static Configuration
+
+### Application Wide
+
+If you want to include an OpenID provider for every application, you could use following configuration.
```
Comment (aixigo-ohornischer) [inline docs/companions.md:15]:
```suggestion
If you want to include an OpenID provider for every application, you could use the following configuration:
```

File: docs/companions.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+"com.github.prevant" = "bar-{{application.name}}"
+```
+
+#### Template Variables
+
+The list of available handlebars variables:
+
+- `application`: The companion's application information
+  - `name`: The application name
+- `services`: An array of the services of the application. Each element has following structure:
```
Comment (aixigo-ohornischer) [inline docs/companions.md:52]:
```suggestion
- `services`: An array of the services of the application. Each element has the following structure:
```

File: docs/companions.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  - `name`: The service name which is equivalent to the network alias
+  - `port`: The exposed port of the service
+  - `type`: The type of service. For example, `instance`, `replica`, `app-companion`, or `service-companion`.
+
+#### Handlebar Helpers
+
+PREvant provides some handlebars helpers which can be used to generate more complex configuration files. See handlerbar's [block helper documentation](https://handlebarsjs.com/block_helpers.html) for more details.
+
+- `{{#isCompanion <type>}}` A conditional handlerbars block helper that checks if the given service type matches any companion type.
+- `isNotCompanion <type>` A conditional handlerbars block helper that checks if the given service type does not match any companion type.
```
Comment (aixigo-ohornischer) [inline docs/companions.md:62]:
```suggestion
PREvant provides some handlebars helpers which can be used to generate more complex configuration files. See handlebar's [block helper documentation](https://handlebarsjs.com/block_helpers.html) for more details.

- `{{#isCompanion <type>}}` A conditional handlebars block helper that checks if the given service type matches any companion type.
- `isNotCompanion <type>` A conditional handlebars block helper that checks if the given service type does not match any companion type.
```

File: docs/companions.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+### Service Based
+
+The service-based companions works the in the same way as the application-based services. Make sure, that the `serviceName` is unique by using the handlebars templating.
```
Comment (aixigo-ohornischer) [inline docs/companions.md:66]:
```suggestion
The service-based companions work the in the same way as the application-based services. Make sure, that the `serviceName` is unique by using handlebars templating.
```

File: docs/companions.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+```
+
+
+#### Template Variables
+
+The list of available handlebars variables:
+
+- `application`: The companion's application information
+  - `name`: The application name
+- `service`: The companion's service containing following fields:
```
Comment (aixigo-ohornischer) [inline docs/companions.md:87]:
```suggestion
- `service`: The companion's service containing the following fields:
```

File: docs/companions.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+Companions offer different deployment strategies so that a companion could be restarted or not under certain conditions. Therefore, PREvant offers following configuration flags:
+
+```toml
+[companions.openid]
+type = 'application'
+image = 'private.example.com/library/openid:latest'
+deploymentStrategy = 'redeploy-on-image-update'
+```
+
+`deploymentStrategy` offers following values and if a companion exists for an app following strategy will be applied:
```
Comment (aixigo-ohornischer) [inline docs/companions.md:103]:
```suggestion
`deploymentStrategy` offers the following values and if a companion exists for an app the following strategy will be applied:
```

File: docs/companions.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+Companions may have varying storage requirements and storage strategies cater to these by offering the below configuration flags:
+
+```toml
+[companions.postgres]
+type = 'application'
+image = 'postgres:latest'
+storageStrategy = 'mount-declared-image-volumes'
+```
+
+`storageStrategy` offers following values to determine how storage is managed for a companion:
```
Comment (aixigo-ohornischer) [inline docs/companions.md:120]:
```suggestion
`storageStrategy` offers the following values to determine how storage is managed for a companion:
```

File: docs/companions.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+- `none` (_default_): Companion is deployed without persistent storage.
+- `mount-declared-image-volumes`: Mounts the volume paths declared within the image, providing persistent storage for the companion.
+
+## Bootstrapping From the Infrastructure Backend
+
+When the [static configuration](#static-configuration) is insufficient for your
+use case, then PREvant can utilize the underlying infrastructure to bootstrap
+the companion configuration from the stdout of containers that are run once
+within the infrastructure (depicted by following image). PREvant's static
```
Comment (aixigo-ohornischer) [inline docs/companions.md:130]:
```suggestion
within the infrastructure (depicted by following the image). PREvant's static
```

File: docs/companions.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+In the depicted images PREvant will start one or more containers on the
+infrastructure backend that are expected to generate output on standard out
+(stdout) that will be parsed by PREvant that needs to be native to the
+underlying infrastructure.
+
+- When PREvant uses Kubernetes as the infrastructure runtime, the bootstrap
+  containers need to output [Kubernetes manifests][k8s-manifest].
+
+  Make sure to put out YAML that is compatible with 1.1 and 1.2
```
Comment (aixigo-ohornischer) [inline docs/companions.md:146]:
```suggestion
  Make sure to output YAML that is compatible with 1.1 and 1.2
```

File: docs/companions.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+  compose][docker-compose] files.
+
+Then before deploying these bootstrapped companions PREvant merges them with
+the objects generated from the HTTP request payload (all bootstrapped
+companions will be considered as application companions). Thus you can add or
+overwrite configurations. For example, you can change the image used or an
+environment variable. If you overwrite any configuration the companion will be
+turned into an instance (as PREvant did before).
+
+Following configuration block depicts that an image
```
Comment (aixigo-ohornischer) [inline docs/companions.md:159]:
```suggestion
The following configuration block depicts that an image
```

File: docs/companions.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+`registry.example.com/user/bootstrap-helm-chart:lastet` is based on a [Helm
+chart][helm-chart] that generates an OpenID provider for your application with
+[Keycloak][keycloak]. Additionally, it is possible to passed additional
```
Comment (aixigo-ohornischer) [inline docs/companions.md:162]:
```suggestion
[Keycloak][keycloak]. Additionally, it is possible to pass additional
```

**Comment by schrieveslaach** [comment]
@aixigo-ohornischer, thanks for proofing this PR.

File: docs/companions.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+the objects generated from the HTTP request payload (all bootstrapped
+companions will be considered as application companions). Thus you can add or
+overwrite configurations. For example, you can change the image used or an
+environment variable. If you overwrite any configuration the companion will be
+turned into an instance (as PREvant did before).
+
+The following configuration block depicts that an image
+`registry.example.com/user/bootstrap-helm-chart:lastet` is based on a [Helm
+chart][helm-chart] that generates an OpenID provider for your application with
+[Keycloak][keycloak]. Additionally, it is possible to pasadditional arguments
```
Comment (samuchila) [inline docs/companions.md:167]:
```suggestion
[Keycloak][keycloak]. Additionally, it is possible to pass additional arguments
```

File: api/src/infrastructure/kubernetes/deployment_unit.rs [hunk 16]
```diff
@@ -0,0 +1,1195 @@
...
+                    "tls.key": "RXhhbXBsZSBkYXRhIGZvciB0aGUgVExTIGNydCBmaWVsZA=="
+                }
+            }])
+        )
+    }
+
+    #[tokio::test]
+    async fn parse_unit_from_deploymen_stdout() {
```
Comment (samuchila) [inline api/src/infrastructure/kubernetes/deployment_unit.rs:903]:
```suggestion
    async fn parse_unit_from_deployment_stdout() {
```

File: api/src/infrastructure/kubernetes/infrastructure.rs [hunk 17]
```diff
@@ -308,163 +301,49 @@ impl KubernetesInfrastructure {
...
+            registries_and_credentials,
+        ))
     }
 
-    async fn deploy_service<'a>(
+    async fn create_payloads(
         &self,
         app_name: &AppName,
-        service: &'a DeployableService,
+        deployabel_service: &DeployableService,
```
Comment (samuchila) [inline api/src/infrastructure/kubernetes/infrastructure.rs:316]:
```suggestion
        deployable_service: &DeployableService,
```

File: api/src/infrastructure/kubernetes/payloads.rs [hunk 18]
```diff
@@ -159,6 +170,134 @@ impl TryFrom<IngressRoute> for TraefikIngressRoute {
...
+            kind: Some(String::from("Service")),
+            name: path.backend.service.clone().unwrap().name,
+            port: Some(
+                path.backend
+                    .service
+                    .as_ref()
+                    .and_then(|service| service.port.as_ref())
+                    .and_then(|port| port.number)
+                    .map(|p| p as u16)
+                    // TODO: how to get the if missing
```
Comment (samuchila) [inline api/src/infrastructure/kubernetes/payloads.rs:281]: I think this comment is incomplete here. Please expand on it for future tracking

**Review by samuchila — CHANGES_REQUESTED** [review]
Some minor typo's identified after reading the changes

File: docs/companions.md [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+companions will be considered as application companions). Thus you can add or
+overwrite configurations. For example, you can change the image used or an
+environment variable. If you overwrite any configuration the companion will be
+turned into an instance (as PREvant did before).
+
+The following configuration block depicts that an image
+`registry.example.com/user/bootstrap-helm-chart:lastet` is based on a [Helm
+chart][helm-chart] that generates an OpenID provider for your application with
+[Keycloak][keycloak]. Additionally, it is possible to pass additional arguments
+to the container than can be templated with [Handlebars][handlebars].
```
Comment (AnumDastgir) [inline docs/companions.md:168]: to the container (than->that) can be templated with [Handlebars][handlebars]

**Pull request merged**